ICE: eei on impl: `expected statement`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_expand/src/base.rs:176:18: 'expected statement'', 'thread 'rustc' ($TID) panicked at compiler/rustc_expand/src/base.rs:176:18: 'expected statement''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
fn main() {
    #[eii]
    impl Bar {}
}

````

original:
````rust
#![feature(extern_item_impls)]
fn main() {
    #[eii]
    impl Bar {}
}

````

Version information
````
rustc 1.94.0-nightly (3f4dc1e02 2025-12-14)
binary: rustc
commit-hash: 3f4dc1e02d759aa3c3807d4efc1f7f6e293536a5
commit-date: 2025-12-14
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.5

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/3f4dc1e02d759aa3c3807d4efc1f7f6e293536a5/compiler/rustc_expand/src/base.rs#L170-L182

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: use of unstable library feature `extern_item_impls`
 --> /tmp/icemaker_global_tempdir.H7kAaxdihhDG/rustc_testrunner_tmpdir_reporting.UmhLa2FjaPWt/mvce.rs:2:7
  |
2 |     #[eii]
  |       ^^^
  |
  = note: see issue #125418 <https://github.com/rust-lang/rust/issues/125418> for more information
  = help: add `#![feature(extern_item_impls)]` to the crate attributes to enable
  = note: this compiler was built on 2025-12-14; consider upgrading it if it is out of date

error: `#[eii]` is only valid on functions
 --> /tmp/icemaker_global_tempdir.H7kAaxdihhDG/rustc_testrunner_tmpdir_reporting.UmhLa2FjaPWt/mvce.rs:2:5
  |
2 |     #[eii]
  |     ^^^^^^


thread 'rustc' (3711055) panicked at compiler/rustc_expand/src/base.rs:176:18:
expected statement
stack backtrace:
   0:     0x7f6b52190173 - <<std[e49fa185a119b89b]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[17865207963548b5]::fmt::Display>::fmt
   1:     0x7f6b52810648 - core[17865207963548b5]::fmt::write
   2:     0x7f6b521a67c6 - <std[e49fa185a119b89b]::sys::stdio::unix::Stderr as std[e49fa185a119b89b]::io::Write>::write_fmt
   3:     0x7f6b52166938 - std[e49fa185a119b89b]::panicking::default_hook::{closure#0}
   4:     0x7f6b52184043 - std[e49fa185a119b89b]::panicking::default_hook
   5:     0x7f6b511741d7 - std[e49fa185a119b89b]::panicking::update_hook::<alloc[756b9f94e2773738]::boxed::Box<rustc_driver_impl[e1c4671be5271ce8]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f6b52184322 - std[e49fa185a119b89b]::panicking::panic_with_hook
   7:     0x7f6b52166a2a - std[e49fa185a119b89b]::panicking::panic_handler::{closure#0}
   8:     0x7f6b5215dab9 - std[e49fa185a119b89b]::sys::backtrace::__rust_end_short_backtrace::<std[e49fa185a119b89b]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f6b521685ad - __rustc[df3bb69314ad135f]::rust_begin_unwind
  10:     0x7f6b4f46df0c - core[17865207963548b5]::panicking::panic_fmt
  11:     0x7f6b53db47a5 - <rustc_expand[ad9a400ec1ef25e2]::expand::MacroExpander>::fully_expand_fragment
  12:     0x7f6b53d5f998 - <rustc_expand[ad9a400ec1ef25e2]::expand::MacroExpander>::expand_crate
  13:     0x7f6b52ff45b2 - rustc_interface[50af632e35c6bce3]::passes::configure_and_expand
  14:     0x7f6b53965e98 - rustc_interface[50af632e35c6bce3]::passes::resolver_for_lowering_raw
  15:     0x7f6b53965c0d - rustc_query_impl[462c6a5584cea8d4]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[462c6a5584cea8d4]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7055997c00c0bb98]::query::erase::Erased<[u8; 16usize]>>
  16:     0x7f6b53965be7 - <rustc_query_impl[462c6a5584cea8d4]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[17865207963548b5]::ops::function::FnOnce<(rustc_middle[7055997c00c0bb98]::ty::context::TyCtxt, ())>>::call_once
  17:     0x7f6b53971b1f - rustc_query_system[5be0c069f6d5c539]::query::plumbing::try_execute_query::<rustc_query_impl[462c6a5584cea8d4]::DynamicConfig<rustc_query_system[5be0c069f6d5c539]::query::caches::SingleCache<rustc_middle[7055997c00c0bb98]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[462c6a5584cea8d4]::plumbing::QueryCtxt, false>
  18:     0x7f6b5397167c - rustc_query_impl[462c6a5584cea8d4]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  19:     0x7f6b53b2ccfb - <rustc_interface[50af632e35c6bce3]::passes::create_and_enter_global_ctxt<core[17865207963548b5]::option::Option<rustc_interface[50af632e35c6bce3]::queries::Linker>, rustc_driver_impl[e1c4671be5271ce8]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[17865207963548b5]::ops::function::FnOnce<(&rustc_session[307f8e7a4263bd4]::session::Session, rustc_middle[7055997c00c0bb98]::ty::context::CurrentGcx, alloc[756b9f94e2773738]::sync::Arc<rustc_data_structures[30161969bbcb553f]::jobserver::Proxy>, &std[e49fa185a119b89b]::sync::once_lock::OnceLock<rustc_middle[7055997c00c0bb98]::ty::context::GlobalCtxt>, &rustc_data_structures[30161969bbcb553f]::sync::worker_local::WorkerLocal<rustc_middle[7055997c00c0bb98]::arena::Arena>, &rustc_data_structures[30161969bbcb553f]::sync::worker_local::WorkerLocal<rustc_hir[5d00f225b65c0257]::Arena>, rustc_driver_impl[e1c4671be5271ce8]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  20:     0x7f6b539c19da - rustc_interface[50af632e35c6bce3]::interface::run_compiler::<(), rustc_driver_impl[e1c4671be5271ce8]::run_compiler::{closure#0}>::{closure#1}
  21:     0x7f6b53931d8e - std[e49fa185a119b89b]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[50af632e35c6bce3]::util::run_in_thread_with_globals<rustc_interface[50af632e35c6bce3]::util::run_in_thread_pool_with_globals<rustc_interface[50af632e35c6bce3]::interface::run_compiler<(), rustc_driver_impl[e1c4671be5271ce8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  22:     0x7f6b53931b60 - <std[e49fa185a119b89b]::thread::lifecycle::spawn_unchecked<rustc_interface[50af632e35c6bce3]::util::run_in_thread_with_globals<rustc_interface[50af632e35c6bce3]::util::run_in_thread_pool_with_globals<rustc_interface[50af632e35c6bce3]::interface::run_compiler<(), rustc_driver_impl[e1c4671be5271ce8]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[17865207963548b5]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  23:     0x7f6b539374f8 - <std[e49fa185a119b89b]::sys::thread::unix::Thread>::new::thread_start
  24:     0x7f6b4d2969cb - <unknown>
  25:     0x7f6b4d31aa0c - <unknown>
  26:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.94.0-nightly (3f4dc1e02 2025-12-14) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [resolver_for_lowering_raw] getting the resolver for lowering

-->
@rustbot label +F-extern_item_impls

--- a/compiler/rustc_builtin_macros/src/eii.rs
+++ b/compiler/rustc_builtin_macros/src/eii.rs
@@ -60,12 +60,18 @@ fn eii_(
 ) -> Vec<Annotatable> {
     let eii_attr_span = ecx.with_def_site_ctxt(eii_attr_span);
 
-    let (item, stmt) = if let Annotatable::Item(item) = item {
-        (item, false)
+    let (item, wrap_item): (_, &dyn Fn(_) -> _) = if let Annotatable::Item(item) = item {
+        (item, &Annotatable::Item)
     } else if let Annotatable::Stmt(ref stmt) = item
         && let StmtKind::Item(ref item) = stmt.kind
     {
-        (item.clone(), true)
+        (item.clone(), &|item| {
+            Annotatable::Stmt(Box::new(Stmt {
+                id: DUMMY_NODE_ID,
+                kind: StmtKind::Item(item),
+                span: eii_attr_span,
+            }))
+        })
     } else {
         ecx.dcx().emit_err(EiiSharedMacroExpectedFunction {
             span: eii_attr_span,
@@ -74,23 +80,25 @@ fn eii_(
         return vec![item];
     };
 
-    let orig_item = item.clone();
-
-    let item = *item;
-
-    let ast::Item { attrs, id: _, span: _, vis, kind: ItemKind::Fn(func), tokens: _ } = item else {
+    let ast::Item { attrs, id: _, span: _, vis, kind: ItemKind::Fn(func), tokens: _ } =
+        item.as_ref()
+    else {
         ecx.dcx().emit_err(EiiSharedMacroExpectedFunction {
             span: eii_attr_span,
             name: path_to_string(&meta_item.path),
         });
-        return vec![Annotatable::Item(Box::new(item))];
+        return vec![wrap_item(item)];
     };
+    // only clone what we need
+    let attrs = attrs.clone();
+    let func = (**func).clone();
+    let vis = vis.clone();
 
     let attrs_from_decl =
         filter_attrs_for_multiple_eii_attr(ecx, attrs, eii_attr_span, &meta_item.path);
 
     let Ok(macro_name) = name_for_impl_macro(ecx, &func, &meta_item) else {
-        return vec![Annotatable::Item(orig_item)];
+        return vec![wrap_item(item)];
     };
 
     // span of the declaring item without attributes
@@ -115,7 +123,7 @@ fn eii_(
         ecx,
         eii_attr_span,
         item_span,
-        *func,
+        func,
         vis,
         &attrs_from_decl,
     )));
@@ -128,20 +136,7 @@ fn eii_(
         decl_span,
     )));
 
-    if stmt {
-        return_items
-            .into_iter()
-            .map(|i| {
-                Annotatable::Stmt(Box::new(Stmt {
-                    id: DUMMY_NODE_ID,
-                    kind: StmtKind::Item(i),
-                    span: eii_attr_span,
-                }))
-            })
-            .collect()
-    } else {
-        return_items.into_iter().map(|i| Annotatable::Item(i)).collect()
-    }
+    return_items.into_iter().map(wrap_item).collect()
 }
 
 /// Decide on the name of the macro that can be used to implement the EII.

--- a/compiler/rustc_builtin_macros/src/eii.rs
+++ b/compiler/rustc_builtin_macros/src/eii.rs
@@ -1,11 +1,12 @@
 use rustc_ast::token::{Delimiter, TokenKind};
 use rustc_ast::tokenstream::{DelimSpacing, DelimSpan, Spacing, TokenStream, TokenTree};
 use rustc_ast::{
-    DUMMY_NODE_ID, EiiExternTarget, EiiImpl, ItemKind, Stmt, StmtKind, ast, token, tokenstream,
+    Attribute, DUMMY_NODE_ID, EiiExternTarget, EiiImpl, ItemKind, MetaItem, Path, Stmt, StmtKind,
+    Visibility, ast,
 };
 use rustc_ast_pretty::pprust::path_to_string;
 use rustc_expand::base::{Annotatable, ExtCtxt};
-use rustc_span::{Ident, Span, kw, sym};
+use rustc_span::{ErrorGuaranteed, Ident, Span, kw, sym};
 use thin_vec::{ThinVec, thin_vec};
 
 use crate::errors::{
@@ -52,12 +53,12 @@ pub(crate) fn unsafe_eii(
 
 fn eii_(
     ecx: &mut ExtCtxt<'_>,
-    span: Span,
+    eii_attr_span: Span,
     meta_item: &ast::MetaItem,
     item: Annotatable,
     impl_unsafe: bool,
 ) -> Vec<Annotatable> {
-    let span = ecx.with_def_site_ctxt(span);
+    let eii_attr_span = ecx.with_def_site_ctxt(eii_attr_span);
 
     let (item, stmt) = if let Annotatable::Item(item) = item {
         (item, false)
@@ -67,7 +68,7 @@ fn eii_(
         (item.clone(), true)
     } else {
         ecx.dcx().emit_err(EiiSharedMacroExpectedFunction {
-            span,
+            span: eii_attr_span,
             name: path_to_string(&meta_item.path),
         });
         return vec![item];
@@ -77,113 +78,218 @@ fn eii_(
 
     let item = *item;
 
-    let ast::Item { attrs, id: _, span: item_span, vis, kind: ItemKind::Fn(mut func), tokens: _ } =
-        item
-    else {
+    let ast::Item { attrs, id: _, span: _, vis, kind: ItemKind::Fn(func), tokens: _ } = item else {
         ecx.dcx().emit_err(EiiSharedMacroExpectedFunction {
-            span,
+            span: eii_attr_span,
             name: path_to_string(&meta_item.path),
         });
         return vec![Annotatable::Item(Box::new(item))];
     };
 
-    // Detect when this is the *second* eii attribute on an item.
-    let mut new_attrs = ThinVec::new();
-    for i in attrs {
-        if i.has_name(sym::eii) {
-            ecx.dcx().emit_err(EiiOnlyOnce {
-                span: i.span,
-                first_span: span,
-                name: path_to_string(&meta_item.path),
-            });
-        } else {
-            new_attrs.push(i);
-        }
-    }
-    let attrs = new_attrs;
+    let attrs_from_decl =
+        filter_attrs_for_multiple_eii_attr(ecx, attrs, eii_attr_span, &meta_item.path);
 
-    let macro_name = if meta_item.is_word() {
-        func.ident
-    } else if let Some([first]) = meta_item.meta_item_list()
-        && let Some(m) = first.meta_item()
-        && m.path.segments.len() == 1
-    {
-        m.path.segments[0].ident
-    } else {
-        ecx.dcx().emit_err(EiiMacroExpectedMaxOneArgument {
-            span: meta_item.span,
-            name: path_to_string(&meta_item.path),
-        });
+    let Ok(macro_name) = name_for_impl_macro(ecx, &func, &meta_item) else {
         return vec![Annotatable::Item(orig_item)];
     };
+
+    // span of the declaring item without attributes
+    let item_span = func.sig.span;
+    // span of the eii attribute and the item below it, i.e. the full declaration
+    let decl_span = eii_attr_span.to(item_span);
+    let foreign_item_name = func.ident;
 
     let mut return_items = Vec::new();
 
     if func.body.is_some() {
-        let mut default_func = func.clone();
-        func.body = None;
-        default_func.eii_impls.push(ast::EiiImpl {
-            node_id: DUMMY_NODE_ID,
-            eii_macro_path: ast::Path::from_ident(macro_name),
-            impl_safety: if impl_unsafe { ast::Safety::Unsafe(span) } else { ast::Safety::Default },
-            span,
-            inner_span: macro_name.span,
-            is_default: true, // important!
-        });
-
-        return_items.push(Box::new(ast::Item {
-            attrs: ThinVec::new(),
-            id: ast::DUMMY_NODE_ID,
-            span,
-            vis: ast::Visibility { span, kind: ast::VisibilityKind::Inherited, tokens: None },
-            kind: ast::ItemKind::Const(Box::new(ast::ConstItem {
-                ident: Ident { name: kw::Underscore, span },
-                defaultness: ast::Defaultness::Final,
-                generics: ast::Generics::default(),
-                ty: Box::new(ast::Ty {
-                    id: DUMMY_NODE_ID,
-                    kind: ast::TyKind::Tup(ThinVec::new()),
-                    span,
-                    tokens: None,
-                }),
-                rhs: Some(ast::ConstItemRhs::Body(Box::new(ast::Expr {
-                    id: DUMMY_NODE_ID,
-                    kind: ast::ExprKind::Block(
-                        Box::new(ast::Block {
-                            stmts: thin_vec![ast::Stmt {
-                                id: DUMMY_NODE_ID,
-                                kind: ast::StmtKind::Item(Box::new(ast::Item {
-                                    attrs: thin_vec![], // FIXME: re-add some original attrs
-                                    id: DUMMY_NODE_ID,
-                                    span: item_span,
-                                    vis: ast::Visibility {
-                                        span,
-                                        kind: ast::VisibilityKind::Inherited,
-                                        tokens: None
-                                    },
-                                    kind: ItemKind::Fn(default_func),
-                                    tokens: None,
-                                })),
-                                span
-                            }],
-                            id: DUMMY_NODE_ID,
-                            rules: ast::BlockCheckMode::Default,
-                            span,
-                            tokens: None,
-                        }),
-                        None,
-                    ),
-                    span,
-                    attrs: ThinVec::new(),
-                    tokens: None,
-                }))),
-                define_opaque: None,
-            })),
-            tokens: None,
-        }))
+        return_items.push(Box::new(generate_default_impl(
+            &func,
+            impl_unsafe,
+            macro_name,
+            eii_attr_span,
+            item_span,
+        )))
     }
 
-    let decl_span = span.to(func.sig.span);
+    return_items.push(Box::new(generate_foreign_item(
+        ecx,
+        eii_attr_span,
+        item_span,
+        *func,
+        vis,
+        &attrs_from_decl,
+    )));
+    return_items.push(Box::new(generate_attribute_macro_to_implement(
+        ecx,
+        eii_attr_span,
+        macro_name,
+        foreign_item_name,
+        impl_unsafe,
+        decl_span,
+    )));
+
+    if stmt {
+        return_items
+            .into_iter()
+            .map(|i| {
+                Annotatable::Stmt(Box::new(Stmt {
+                    id: DUMMY_NODE_ID,
+                    kind: StmtKind::Item(i),
+                    span: eii_attr_span,
+                }))
+            })
+            .collect()
+    } else {
+        return_items.into_iter().map(|i| Annotatable::Item(i)).collect()
+    }
+}
+
+/// Decide on the name of the macro that can be used to implement the EII.
+/// This is either an explicitly given name, or the name of the item in the
+/// declaration of the EII.
+fn name_for_impl_macro(
+    ecx: &mut ExtCtxt<'_>,
+    func: &ast::Fn,
+    meta_item: &MetaItem,
+) -> Result<Ident, ErrorGuaranteed> {
+    if meta_item.is_word() {
+        Ok(func.ident)
+    } else if let Some([first]) = meta_item.meta_item_list()
+        && let Some(m) = first.meta_item()
+        && m.path.segments.len() == 1
+    {
+        Ok(m.path.segments[0].ident)
+    } else {
+        Err(ecx.dcx().emit_err(EiiMacroExpectedMaxOneArgument {
+            span: meta_item.span,
+            name: path_to_string(&meta_item.path),
+        }))
+    }
+}
+
+/// Ensure that in the list of attrs, there's only a single `eii` attribute.
+fn filter_attrs_for_multiple_eii_attr(
+    ecx: &mut ExtCtxt<'_>,
+    attrs: ThinVec<Attribute>,
+    eii_attr_span: Span,
+    eii_attr_path: &Path,
+) -> ThinVec<Attribute> {
+    attrs
+        .into_iter()
+        .filter(|i| {
+            if i.has_name(sym::eii) {
+                ecx.dcx().emit_err(EiiOnlyOnce {
+                    span: i.span,
+                    first_span: eii_attr_span,
+                    name: path_to_string(eii_attr_path),
+                });
+                false
+            } else {
+                true
+            }
+        })
+        .collect()
+}
+
+fn generate_default_impl(
+    func: &ast::Fn,
+    impl_unsafe: bool,
+    macro_name: Ident,
+    eii_attr_span: Span,
+    item_span: Span,
+) -> ast::Item {
+    // FIXME: re-add some original attrs
+    let attrs = ThinVec::new();
+
+    let mut default_func = func.clone();
+    default_func.eii_impls.push(EiiImpl {
+        node_id: DUMMY_NODE_ID,
+        inner_span: macro_name.span,
+        eii_macro_path: ast::Path::from_ident(macro_name),
+        impl_safety: if impl_unsafe {
+            ast::Safety::Unsafe(eii_attr_span)
+        } else {
+            ast::Safety::Default
+        },
+        span: eii_attr_span,
+        is_default: true,
+    });
+
+    ast::Item {
+        attrs: ThinVec::new(),
+        id: ast::DUMMY_NODE_ID,
+        span: eii_attr_span,
+        vis: ast::Visibility {
+            span: eii_attr_span,
+            kind: ast::VisibilityKind::Inherited,
+            tokens: None,
+        },
+        kind: ast::ItemKind::Const(Box::new(ast::ConstItem {
+            ident: Ident { name: kw::Underscore, span: eii_attr_span },
+            defaultness: ast::Defaultness::Final,
+            generics: ast::Generics::default(),
+            ty: Box::new(ast::Ty {
+                id: DUMMY_NODE_ID,
+                kind: ast::TyKind::Tup(ThinVec::new()),
+                span: eii_attr_span,
+                tokens: None,
+            }),
+            rhs: Some(ast::ConstItemRhs::Body(Box::new(ast::Expr {
+                id: DUMMY_NODE_ID,
+                kind: ast::ExprKind::Block(
+                    Box::new(ast::Block {
+                        stmts: thin_vec![ast::Stmt {
+                            id: DUMMY_NODE_ID,
+                            kind: ast::StmtKind::Item(Box::new(ast::Item {
+                                attrs,
+                                id: DUMMY_NODE_ID,
+                                span: item_span,
+                                vis: ast::Visibility {
+                                    span: eii_attr_span,
+                                    kind: ast::VisibilityKind::Inherited,
+                                    tokens: None
+                                },
+                                kind: ItemKind::Fn(Box::new(default_func)),
+                                tokens: None,
+                            })),
+                            span: eii_attr_span
+                        }],
+                        id: DUMMY_NODE_ID,
+                        rules: ast::BlockCheckMode::Default,
+                        span: eii_attr_span,
+                        tokens: None,
+                    }),
+                    None,
+                ),
+                span: eii_attr_span,
+                attrs: ThinVec::new(),
+                tokens: None,
+            }))),
+            define_opaque: None,
+        })),
+        tokens: None,
+    }
+}
+
+/// Generates a foreign item, like
+///
+/// ```rust, ignore
+/// extern "…" { safe fn item(); }
+/// ```
+fn generate_foreign_item(
+    ecx: &mut ExtCtxt<'_>,
+    eii_attr_span: Span,
+    item_span: Span,
+    mut func: ast::Fn,
+    vis: Visibility,
+    attrs_from_decl: &[Attribute],
+) -> ast::Item {
+    let mut foreign_item_attrs = ThinVec::new();
+    foreign_item_attrs.extend_from_slice(attrs_from_decl);
+
+    // Add the rustc_eii_extern_item on the foreign item. Usually, foreign items are mangled.
+    // This attribute makes sure that we later know that this foreign item's symbol should not be.
+    foreign_item_attrs.push(ecx.attr_word(sym::rustc_eii_extern_item, eii_attr_span));
 
     let abi = match func.sig.header.ext {
         // extern "X" fn  =>  extern "X" {}
@@ -196,85 +302,69 @@ fn eii_(
             suffix: None,
             symbol_unescaped: sym::Rust,
             style: ast::StrStyle::Cooked,
-            span,
+            span: eii_attr_span,
         }),
     };
 
     // ABI has been moved to the extern {} block, so we remove it from the fn item.
     func.sig.header.ext = ast::Extern::None;
+    func.body = None;
 
     // And mark safe functions explicitly as `safe fn`.
     if func.sig.header.safety == ast::Safety::Default {
         func.sig.header.safety = ast::Safety::Safe(func.sig.span);
     }
 
-    // extern "…" { safe fn item(); }
-    let mut extern_item_attrs = attrs.clone();
-    extern_item_attrs.push(ast::Attribute {
-        kind: ast::AttrKind::Normal(Box::new(ast::NormalAttr {
-            item: ast::AttrItem {
-                unsafety: ast::Safety::Default,
-                // Add the rustc_eii_extern_item on the foreign item. Usually, foreign items are mangled.
-                // This attribute makes sure that we later know that this foreign item's symbol should not be.
-                path: ast::Path::from_ident(Ident::new(sym::rustc_eii_extern_item, span)),
-                args: ast::AttrArgs::Empty,
-                tokens: None,
-            },
-            tokens: None,
-        })),
-        id: ecx.sess.psess.attr_id_generator.mk_attr_id(),
-        style: ast::AttrStyle::Outer,
-        span,
-    });
-
-    let extern_block = Box::new(ast::Item {
+    ast::Item {
         attrs: ast::AttrVec::default(),
         id: ast::DUMMY_NODE_ID,
-        span,
-        vis: ast::Visibility { span, kind: ast::VisibilityKind::Inherited, tokens: None },
+        span: eii_attr_span,
+        vis: ast::Visibility {
+            span: eii_attr_span,
+            kind: ast::VisibilityKind::Inherited,
+            tokens: None,
+        },
         kind: ast::ItemKind::ForeignMod(ast::ForeignMod {
-            extern_span: span,
-            safety: ast::Safety::Unsafe(span),
+            extern_span: eii_attr_span,
+            safety: ast::Safety::Unsafe(eii_attr_span),
             abi,
             items: From::from([Box::new(ast::ForeignItem {
-                attrs: extern_item_attrs,
+                attrs: foreign_item_attrs,
                 id: ast::DUMMY_NODE_ID,
                 span: item_span,
                 vis,
-                kind: ast::ForeignItemKind::Fn(func.clone()),
+                kind: ast::ForeignItemKind::Fn(Box::new(func.clone())),
                 tokens: None,
             })]),
         }),
         tokens: None,
-    });
+    }
+}
 
-    let mut macro_attrs = attrs.clone();
-    macro_attrs.push(
-        // #[builtin_macro(eii_shared_macro)]
-        ast::Attribute {
-            kind: ast::AttrKind::Normal(Box::new(ast::NormalAttr {
-                item: ast::AttrItem {
-                    unsafety: ast::Safety::Default,
-                    path: ast::Path::from_ident(Ident::new(sym::rustc_builtin_macro, span)),
-                    args: ast::AttrArgs::Delimited(ast::DelimArgs {
-                        dspan: DelimSpan::from_single(span),
-                        delim: Delimiter::Parenthesis,
-                        tokens: TokenStream::new(vec![tokenstream::TokenTree::token_alone(
-                            token::TokenKind::Ident(sym::eii_shared_macro, token::IdentIsRaw::No),
-                            span,
-                        )]),
-                    }),
-                    tokens: None,
-                },
-                tokens: None,
-            })),
-            id: ecx.sess.psess.attr_id_generator.mk_attr_id(),
-            style: ast::AttrStyle::Outer,
-            span,
-        },
-    );
+/// Generate a stub macro (a bit like in core) that will roughly look like:
+///
+/// ```rust, ignore, example
+/// // Since this a stub macro, the actual code that expands it lives in the compiler.
+/// // This attribute tells the compiler that
+/// #[builtin_macro(eii_shared_macro)]
+/// // the metadata to link this macro to the generated foreign item.
+/// #[eii_extern_target(<related_reign_item>)]
+/// macro macro_name { () => {} }
+/// ```
+fn generate_attribute_macro_to_implement(
+    ecx: &mut ExtCtxt<'_>,
+    span: Span,
+    macro_name: Ident,
+    foreign_item_name: Ident,
+    impl_unsafe: bool,
+    decl_span: Span,
+) -> ast::Item {
+    let mut macro_attrs = ThinVec::new();
 
-    let macro_def = Box::new(ast::Item {
+    // #[builtin_macro(eii_shared_macro)]
+    macro_attrs.push(ecx.attr_nested_word(sym::rustc_builtin_macro, sym::eii_shared_macro, span));
+
+    ast::Item {
         attrs: macro_attrs,
         id: ast::DUMMY_NODE_ID,
         span,
@@ -305,33 +395,15 @@ fn eii_(
                     ]),
                 }),
                 macro_rules: false,
-                // #[eii_extern_target(func.ident)]
+                // #[eii_extern_target(foreign_item_ident)]
                 eii_extern_target: Some(ast::EiiExternTarget {
-                    extern_item_path: ast::Path::from_ident(func.ident),
+                    extern_item_path: ast::Path::from_ident(foreign_item_name),
                     impl_unsafe,
                     span: decl_span,
                 }),
             },
         ),
         tokens: None,
-    });
-
-    return_items.push(extern_block);
-    return_items.push(macro_def);
-
-    if stmt {
-        return_items
-            .into_iter()
-            .map(|i| {
-                Annotatable::Stmt(Box::new(Stmt {
-                    id: DUMMY_NODE_ID,
-                    kind: StmtKind::Item(i),
-                    span,
-                }))
-            })
-            .collect()
-    } else {
-        return_items.into_iter().map(|i| Annotatable::Item(i)).collect()
     }
 }
 
@@ -436,10 +508,10 @@ pub(crate) fn eii_shared_macro(
 
     f.eii_impls.push(EiiImpl {
         node_id: DUMMY_NODE_ID,
+        inner_span: meta_item.path.span,
         eii_macro_path: meta_item.path.clone(),
         impl_safety: meta_item.unsafety,
         span,
-        inner_span: meta_item.path.span,
         is_default,
     });
 

--- a/tests/ui/eii/error_statement_position.rs
+++ b/tests/ui/eii/error_statement_position.rs
@@ -1,6 +1,16 @@
 #![feature(extern_item_impls)]
+// EIIs can, despite not being super useful, be declared in statement position
+// nested inside items. Items in statement position, when expanded as part of a macro,
+// need to be wrapped slightly differently (in an `ast::Statement`).
+// We did this on the happy path (no errors), but when there was an error, we'd
+// replace it with *just* an `ast::Item` not wrapped in an `ast::Statement`.
+// This caused an ICE (https://github.com/rust-lang/rust/issues/149980).
+// this test fails to build, but demonstrates that no ICE is produced.
 
 fn main() {
+    struct Bar;
+
     #[eii]
+    //~^ ERROR `#[eii]` is only valid on functions
     impl Bar {}
 }

--- a/tests/ui/eii/error_statement_position.rs
+++ b/tests/ui/eii/error_statement_position.rs
@@ -1,0 +1,6 @@
+#![feature(extern_item_impls)]
+
+fn main() {
+    #[eii]
+    impl Bar {}
+}

--- a/tests/ui/eii/error_statement_position.stderr
+++ b/tests/ui/eii/error_statement_position.stderr
@@ -1,0 +1,8 @@
+error: `#[eii]` is only valid on functions
+  --> $DIR/error_statement_position.rs:13:5
+   |
+LL |     #[eii]
+   |     ^^^^^^
+
+error: aborting due to 1 previous error
+

